File: Python/lld-examples/hard/distributed-chat-system/message_service.py
"""
MessageService class for the distributed chat system
"""

from typing import Dict, List, Optional, Set
from datetime import datetime, timedelta
import threading
import time
from collections import defaultdict, deque

from message import Message, MessageType, MessageStatus
from user import User


class MessageService:
    """MessageService class for managing messages"""
    
    def __init__(self):
        # Message storage
        self.messages: Dict[str, Message] = {}  # message_id -> Message
        self.user_messages: Dict[str, List[str]] = defaultdict(list)  # user_id -> [message_ids]
        self.room_messages: Dict[str, List[str]] = defaultdict(list)  # room_id -> [message_ids]
        self.private_messages: Dict[str, List[str]] = defaultdict(list)  # user_id -> [message_ids]
        
        # Message queues for delivery
        self.pending_messages: deque = deque()
        self.failed_messages: List[Message] = []
        
        # Message threading
        self.threads: Dict[str, List[str]] = defaultdict(list)  # thread_id -> [message_ids]
        
        # Message history limits
        self.max_messages_per_room = 10000
        self.max_messages_per_user = 5000
        self.message_retention_days = 30
        
        # Statistics
        self.total_messages_sent = 0
        self.total_messages_delivered = 0
        self.total_messages_failed = 0
        
        # Lock for thread safety
        self._lock = threading.Lock()
        
        print("Message service initialized")
    
    def send_message(self, message: Message) -> bool:
        """Send a message"""
        try:
            with self._lock:
                # Store message
                self.messages[message.message_id] = message
                
                # Add to pending queue for delivery
                self.pending_messages.append(message)
                
                # Update statistics
                self.total_messages_sent += 1
                
                # Add to indexes
                self._index_message(message)
                
                print(f"Message queued for delivery: {message.message_id}")
                return True
                
        except Exception as e:
            print(f"Error sending message: {e}")
            return False
    
    def _index_message(self, message: Message):
        """Index message for quick retrieval"""
        # Index by sender
        self.user_messages[message.sender.user_id].append(message.message_id)
        
        # Index by type
        if message.message_type == MessageType.ROOM_MESSAGE and message.room_id:
            self.room_messages[message.room_id].append(message.message_id)
        elif message.message_type == MessageType.PRIVATE_MESSAGE and message.recipient_id:
            # Index for both sender and recipient
            self.private_messages[message.sender.user_id].append(message.message_id)
            self.private_messages[message.recipient_id].append(message.message_id)
        
        # Thread indexing
        if message.thread_id:
            self.threads[message.thread_id].append(message.message_id)
    
    def process_pending_messages(self):
        """Process pending messages for delivery"""
        processed_count = 0
        
        with self._lock:
            # Process up to 100 messages per batch
            for _ in range(min(100, len(self.pending_messages))):
                if not self.pending_messages:
                    break
                
                message = self.pending_messages.popleft()
                
                if self._deliver_message(message):
                    message.mark_as_delivered()
                    self.total_messages_delivered += 1
                    processed_count += 1
                else:
                    message.increment_delivery_attempts()
                    
                    if message.can_retry_delivery():
                        # Re-queue for retry
                        self.pending_messages.append(message)
                    else:
                        # Mark as failed
                        message.mark_as_failed()
                        self.failed_messages.append(message)
                        self.total_messages_failed += 1
        
        if processed_count > 0:
            print(f"Processed {processed_count} messages")
    
    def _deliver_message(self, message: Message) -> bool:
        """Attempt to deliver a message"""
        try:
            # Simulate message delivery
            if message.message_type == MessageType.ROOM_MESSAGE:
                return self._deliver_room_message(message)
            elif message.message_type == MessageType.PRIVATE_MESSAGE:
                return self._deliver_private_message(message)
            elif message.message_type == MessageType.SYSTEM_MESSAGE:
                return self._deliver_system_message(message)
            
            return False
            
        except Exception as e:
            print(f"Error delivering message {message.message_id}: {e}")
            return False
    
    def _deliver_room_message(self, message: Message) -> bool:
        """Deliver room message"""
        # In a real system, this would send to all room members
        print(f"Delivering room message {message.message_id} to room {message.room_id}")
        return True
    
    def _deliver_private_message(self, message: Message) -> bool:
        """Deliver private message"""
        # In a real system, this would send to the recipient
        print(f"Delivering private message {message.message_id} to user {message.recipient_id}")
        return True
    
    def _deliver_system_message(self, message: Message) -> bool:
        """Deliver system message"""
        # In a real system, this would broadcast to all users
        print(f"Delivering system message {message.message_id}")
        return True
    
    def get_message(self, message_id: str) -> Optional[Message]:
        """Get message by ID"""
        return self.messages.get(message_id)
    
    def get_user_messages(self, user_id: str, limit: int = 50) -> List[Message]:
        """Get messages for a user"""
        message_ids = self.user_messages.get(user_id, [])
        
        # Get recent messages
        recent_ids = message_ids[-limit:] if limit > 0 else message_ids
        
        messages = []
        for msg_id in recent_ids:
            message = self.messages.get(msg_id)
            if message:
                messages.append(message)
        
        return sorted(messages, key=lambda m: m.timestamp)
    
    def get_room_messages(self, room_id: str, limit: int = 50) -> List[Message]:
        """Get messages for a room"""
        message_ids = self.room_messages.get(room_id, [])
        
        # Get recent messages
        recent_ids = message_ids[-limit:] if limit > 0 else message_ids
        
        messages = []
        for msg_id in recent_ids:
            message = self.messages.get(msg_id)
            if message:
                messages.append(message)
        
        return sorted(messages, key=lambda m: m.timestamp)
    
    def get_private_messages(self, user_id: str, limit: int = 50) -> List[Message]:
        """Get private messages for a user"""
        message_ids = self.private_messages.get(user_id, [])
        
        # Get recent messages
        recent_ids = message_ids[-limit:] if limit > 0 else message_ids
        
        messages = []
        for msg_id in recent_ids:
            message = self.messages.get(msg_id)
            if message and message.message_type == MessageType.PRIVATE_MESSAGE:
                messages.append(message)
        
        return sorted(messages, key=lambda m: m.timestamp)
    
    def get_conversation_messages(self, user1_id: str, user2_id: str, limit: int = 50) -> List[Message]:
        """Get conversation between two users"""
        user1_private = self.get_private_messages(user1_id)
        conversation = []
        
        for message in user1_private:
            if ((message.sender.user_id == user1_id and message.recipient_id == user2_id) or
                (message.sender.user_id == user2_id and message.recipient_id == user1_id)):
                conversation.append(message)
        
        # Sort by timestamp and limit
        conversation.sort(key=lambda m: m.timestamp)
        return conversation[-limit:] if limit > 0 else conversation
    
    def get_thread_messages(self, thread_id: str) -> List[Message]:
        """Get messages in a thread"""
        message_ids = self.threads.get(thread_id, [])
        
        messages = []
        for msg_id in message_ids:
            message = self.messages.get(msg_id)
            if message:
                messages.append(message)
        
        return sorted(messages, key=lambda m: m.timestamp)
    
    def search_messages(self, query: str, user_id: Optional[str] = None, 
                       room_id: Optional[str] = None, limit: int = 50) -> List[Message]:
        """Search messages by content"""
        query = query.lower()
        results = []
        
        for message in self.messages.values():
            # Filter by user or room if specified
            if user_id and message.sender.user_id != user_id:
                continue
            if room_id and message.room_id != room_id:
                continue
            
            # Search in content
            if query in message.content.lower():
                results.append(message)
        
        # Sort by timestamp (most recent first) and limit
        results.sort(key=lambda m: m.timestamp, reverse=True)
        return results[:limit] if limit > 0 else results
    
    def edit_message(self, message_id: str, new_content: str, editor_user_id: str) -> bool:
        """Edit a message"""
        message = self.messages.get(message_id)
        if not message:
            return False
        
        # Check if user can edit this message
        if message.sender.user_id != editor_user_id:
            return False
        
        # Check edit time limit (e.g., 15 minutes)
        edit_deadline = message.timestamp + timedelta(minutes=15)
        if datetime.now() > edit_deadline:
            return False
        
        message.edit_content(new_content)
        print(f"Message {message_id} edited by user {editor_user_id}")
        return True
    
    def delete_message(self, message_id: str, deleter_user_id: str) -> bool:
        """Delete a message"""
        message = self.messages.get(message_id)
        if not message:
            return False
        
        # Check if user can delete this message
        if message.sender.user_id != deleter_user_id:
            return False
        
        message.delete_message()
        print(f"Message {message_id} deleted by user {deleter_user_id}")
        return True
    
    def add_reaction(self, message_id: str, user_id: str, emoji: str) -> bool:
        """Add reaction to a message"""
        message = self.messages.get(message_id)
        if not message:
            return False
        
        message.add_reaction(user_id, emoji)
        print(f"User {user_id} reacted to message {message_id} with {emoji}")
        return True
    
    def remove_reaction(self, message_id: str, user_id: str) -> bool:
        """Remove reaction from a message"""
        message = self.messages.get(message_id)
        if not message:
            return False
        
        message.remove_reaction(user_id)
        print(f"User {user_id} removed reaction from message {message_id}")
        return True
    
    def mark_as_read(self, message_id: str, user_id: str) -> bool:
        """Mark message as read by user"""
        message = self.messages.get(message_id)
        if not message:
            return False
        
        # In a real system, you'd track read status per user
        if message.recipient_id == user_id or message.room_id:
            message.mark_as_read()
            return True
        
        return False
    
    def get_unread_count(self, user_id: str) -> int:
        """Get unread message count for user"""
        count = 0
        
        # Check private messages
        for msg_id in self.private_messages.get(user_id, []):
            message = self.messages.get(msg_id)
            if (message and 
                message.recipient_id == user_id and 
                message.status != MessageStatus.READ):
                count += 1
        
        return count
    
    def cleanup_old_messages(self):
        """Clean up old messages based on retention policy"""
        cutoff_date = datetime.now() - timedelta(days=self.message_retention_days)
        messages_to_delete = []
        
        for message_id, message in self.messages.items():
            if message.timestamp < cutoff_date:
                messages_to_delete.append(message_id)
        
        # Delete old messages
        for message_id in messages_to_delete:
            self._delete_message_completely(message_id)
        
        if messages_to_delete:
            print(f"Cleaned up {len(messages_to_delete)} old messages")
    
    def _delete_message_completely(self, message_id: str):
        """Completely delete a message from all indexes"""
        message = self.messages.get(message_id)
        if not message:
            return
        
        # Remove from all indexes
        self.user_messages[message.sender.user_id].remove(message_id)
        
        if message.message_type == MessageType.ROOM_MESSAGE and message.room_id:
            self.room_messages[message.room_id].remove(message_id)
        elif message.message_type == MessageType.PRIVATE_MESSAGE and message.recipient_id:
            self.private_messages[message.sender.user_id].remove(message_id)
            self.private_messages[message.recipient_id].remove(message_id)
        
        if message.thread_id:
            self.threads[message.thread_id].remove(message_id)
        
        # Remove from main storage
        del self.messages[message_id]
    
    def get_all_messages(self) -> List[Message]:
        """Get all messages (for admin/debugging)"""
        return list(self.messages.values())
    
    def get_statistics(self) -> dict:
        """Get message service statistics"""
        return {
            'total_messages': len(self.messages),
            'total_messages_sent': self.total_messages_sent,
            'total_messages_delivered': self.total_messages_delivered,
            'total_messages_failed': self.total_messages_failed,
            'pending_messages': len(self.pending_messages),
            'failed_messages': len(self.failed_messages),
            'threads_count': len(self.threads),
            'rooms_with_messages': len(self.room_messages),
            'users_with_messages': len(self.user_messages)
        }
    
    def __str__(self):
        return f"MessageService(messages={len(self.messages)}, pending={len(self.pending_messages)})"
    
    def __repr__(self):
        return self.__str__() 

File: JAVA/lld-examples/hard/distributed-chat-system/src/main/java/com/systemdesign/lld/chat/service/MessageService.java
package com.systemdesign.lld.chat.service;

import com.systemdesign.lld.chat.model.Message;
import com.systemdesign.lld.chat.model.User;
import java.time.LocalDateTime;
import java.util.*;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.stream.Collectors;

/**
 * MessageService class for managing messages in the distributed chat system
 */
public class MessageService {
    private final Map<String, Message> messages; // messageId -> Message
    private final Map<String, List<String>> userMessages; // userId -> List of messageIds
    private final Map<String, List<String>> roomMessages; // roomId -> List of messageIds
    private final Map<String, List<String>> privateMessages; // userId -> List of private messageIds
    
    // Message queues for delivery
    private final BlockingQueue<Message> pendingMessages;
    private final List<Message> failedMessages;
    
    // Message threading
    private final Map<String, List<String>> threads; // threadId -> List of messageIds
    
    // Message history limits
    private final int maxMessagesPerRoom = 10000;
    private final int maxMessagesPerUser = 5000;
    private final int messageRetentionDays = 30;
    
    // Statistics
    private long totalMessagesSent;
    private long totalMessagesDelivered;
    private long totalMessagesFailed;
    
    public MessageService() {
        this.messages = new ConcurrentHashMap<>();
        this.userMessages = new ConcurrentHashMap<>();
        this.roomMessages = new ConcurrentHashMap<>();
        this.privateMessages = new ConcurrentHashMap<>();
        this.pendingMessages = new LinkedBlockingQueue<>();
        this.failedMessages = new ArrayList<>();
        this.threads = new ConcurrentHashMap<>();
        
        this.totalMessagesSent = 0;
        this.totalMessagesDelivered = 0;
        this.totalMessagesFailed = 0;
        
        System.out.println("Message service initialized");
    }
    
    /**
     * Send a message
     */
    public synchronized boolean sendMessage(Message message) {
        try {
            // Store message
            messages.put(message.getMessageId(), message);
            
            // Add to pending queue for delivery
            pendingMessages.offer(message);
            
            // Update statistics
            totalMessagesSent++;
            
            // Index message
            indexMessage(message);
            
            System.out.println("Message queued for delivery: " + message.getMessageId());
            return true;
            
        } catch (Exception e) {
            System.err.println("Error sending message: " + e.getMessage());
            return false;
        }
    }
    
    /**
     * Index message for quick retrieval
     */
    private void indexMessage(Message message) {
        // Index by sender
        userMessages.computeIfAbsent(message.getSender().getUserId(), k -> new ArrayList<>())
                   .add(message.getMessageId());
        
        // Index by type
        if (message.isRoomMessage() && message.getRoomId() != null) {
            roomMessages.computeIfAbsent(message.getRoomId(), k -> new ArrayList<>())
                       .add(message.getMessageId());
        } else if (message.isPrivateMessage() && message.getRecipientId() != null) {
            // Index for both sender and recipient
            privateMessages.computeIfAbsent(message.getSender().getUserId(), k -> new ArrayList<>())
                          .add(message.getMessageId());
            privateMessages.computeIfAbsent(message.getRecipientId(), k -> new ArrayList<>())
                          .add(message.getMessageId());
        }
        
        // Thread indexing
        if (message.getThreadId() != null) {
            threads.computeIfAbsent(message.getThreadId(), k -> new ArrayList<>())
                   .add(message.getMessageId());
        }
    }
    
    /**
     * Process pending messages for delivery
     */
    public void processPendingMessages() {
        int processedCount = 0;
        
        // Process up to 100 messages per batch
        for (int i = 0; i < 100 && !pendingMessages.isEmpty(); i++) {
            Message message = pendingMessages.poll();
            if (message == null) break;
            
            if (deliverMessage(message)) {
                message.markAsDelivered();
                totalMessagesDelivered++;
                processedCount++;
            } else {
                message.incrementDeliveryAttempts();
                
                if (message.canRetryDelivery()) {
                    // Re-queue for retry
                    pendingMessages.offer(message);
                } else {
                    // Mark as failed
                    message.markAsFailed();
                    synchronized (failedMessages) {
                        failedMessages.add(message);
                    }
                    totalMessagesFailed++;
                }
            }
        }
        
        if (processedCount > 0) {
            System.out.println("Processed " + processedCount + " messages");
        }
    }
    
    /**
     * Attempt to deliver a message
     */
    private boolean deliverMessage(Message message) {
        try {
            // Simulate message delivery
            if (message.isRoomMessage()) {
                return deliverRoomMessage(message);
            } else if (message.isPrivateMessage()) {
                return deliverPrivateMessage(message);
            } else if (message.isSystemMessage()) {
                return deliverSystemMessage(message);
            }
            
            return false;
            
        } catch (Exception e) {
            System.err.println("Error delivering message " + message.getMessageId() + ": " + e.getMessage());
            return false;
        }
    }
    
    /**
     * Deliver room message
     */
    private boolean deliverRoomMessage(Message message) {
        // In a real system, this would send to all room members
        System.out.println("Delivering room message " + message.getMessageId() + " to room " + message.getRoomId());
        return true;
    }
    
    /**
     * Deliver private message
     */
    private boolean deliverPrivateMessage(Message message) {
        // In a real system, this would send to the recipient
        System.out.println("Delivering private message " + message.getMessageId() + " to user " + message.getRecipientId());
        return true;
    }
    
    /**
     * Deliver system message
     */
    private boolean deliverSystemMessage(Message message) {
        // In a real system, this would broadcast to all users
        System.out.println("Delivering system message " + message.getMessageId());
        return true;
    }
    
    /**
     * Get message by ID
     */
    public Message getMessage(String messageId) {
        return messages.get(messageId);
    }
    
    /**
     * Get messages for a user
     */
    public List<Message> getUserMessages(String userId, int limit) {
        List<String> messageIds = userMessages.get(userId);
        if (messageIds == null) {
            return new ArrayList<>();
        }
        
        // Get recent messages
        List<String> recentIds = messageIds.subList(
            Math.max(0, messageIds.size() - limit), 
            messageIds.size()
        );
        
        return recentIds.stream()
                .map(messages::get)
                .filter(Objects::nonNull)
                .sorted(Comparator.comparing(Message::getTimestamp))
                .collect(Collectors.toList());
    }
    
    /**
     * Get messages for a room
     */
    public List<Message> getRoomMessages(String roomId, int limit) {
        List<String> messageIds = roomMessages.get(roomId);
        if (messageIds == null) {
            return new ArrayList<>();
        }
        
        // Get recent messages
        List<String> recentIds = messageIds.subList(
            Math.max(0, messageIds.size() - limit), 
            messageIds.size()
        );
        
        return recentIds.stream()
                .map(messages::get)
                .filter(Objects::nonNull)
                .sorted(Comparator.comparing(Message::getTimestamp))
                .collect(Collectors.toList());
    }
    
    /**
     * Get private messages for a user
     */
    public List<Message> getPrivateMessages(String userId, int limit) {
        List<String> messageIds = privateMessages.get(userId);
        if (messageIds == null) {
            return new ArrayList<>();
        }
        
        // Get recent messages
        List<String> recentIds = messageIds.subList(
            Math.max(0, messageIds.size() - limit), 
            messageIds.size()
        );
        
        return recentIds.stream()
                .map(messages::get)
                .filter(Objects::nonNull)
                .filter(msg -> msg.getMessageType() == Message.MessageType.PRIVATE_MESSAGE)
                .sorted(Comparator.comparing(Message::getTimestamp))
                .collect(Collectors.toList());
    }
    
    /**
     * Get conversation between two users
     */
    public List<Message> getConversationMessages(String user1Id, String user2Id, int limit) {
        List<Message> user1Private = getPrivateMessages(user1Id, Integer.MAX_VALUE);
        
        return user1Private.stream()
                .filter(msg -> 
                    (msg.getSender().getUserId().equals(user1Id) && msg.getRecipientId().equals(user2Id)) ||
                    (msg.getSender().getUserId().equals(user2Id) && msg.getRecipientId().equals(user1Id))
                )
                .sorted(Comparator.comparing(Message::getTimestamp))
                .limit(limit)
                .collect(Collectors.toList());
    }
    
    /**
     * Get messages in a thread
     */
    public List<Message> getThreadMessages(String threadId) {
        List<String> messageIds = threads.get(threadId);
        if (messageIds == null) {
            return new ArrayList<>();
        }
        
        return messageIds.stream()
                .map(messages::get)
                .filter(Objects::nonNull)
                .sorted(Comparator.comparing(Message::getTimestamp))
                .collect(Collectors.toList());
    }
    
    /**
     * Search messages by content
     */
    public List<Message> searchMessages(String query, String userId, String roomId, int limit) {
        String lowerQuery = query.toLowerCase();
        
        return messages.values().stream()
                .filter(msg -> {
                    // Filter by user or room if specified
                    if (userId != null && !msg.getSender().getUserId().equals(userId)) {
                        return false;
                    }
                    if (roomId != null && !roomId.equals(msg.getRoomId())) {
                        return false;
                    }
                    
                    // Search in content
                    return msg.getContent().toLowerCase().contains(lowerQuery);
                })
                .sorted(Comparator.comparing(Message::getTimestamp).reversed())
                .limit(limit)
                .collect(Collectors.toList());
    }
    
    /**
     * Edit a message
     */
    public boolean editMessage(String messageId, String newContent, String editorUserId) {
        Message message = messages.get(messageId);
        if (message == null) {
            return false;
        }
        
        // Check if user can edit this message
        if (!message.getSender().getUserId().equals(editorUserId)) {
            return false;
        }
        
        // Check edit time limit (e.g., 15 minutes)
        LocalDateTime editDeadline = message.getTimestamp().plusMinutes(15);
        if (LocalDateTime.now().isAfter(editDeadline)) {
            return false;
        }
        
        message.editContent(newContent);
        System.out.println("Message " + messageId + " edited by user " + editorUserId);
        return true;
    }
    
    /**
     * Delete a message
     */
    public boolean deleteMessage(String messageId, String deleterUserId) {
        Message message = messages.get(messageId);
        if (message == null) {
            return false;
        }
        
        // Check if user can delete this message
        if (!message.getSender().getUserId().equals(deleterUserId)) {
            return false;
        }
        
        message.deleteMessage();
        System.out.println("Message " + messageId + " deleted by user " + deleterUserId);
        return true;
    }
    
    /**
     * Add reaction to a message
     */
    public boolean addReaction(String messageId, String userId, String emoji) {
        Message message = messages.get(messageId);
        if (message == null) {
            return false;
        }
        
        message.addReaction(userId, emoji);
        System.out.println("User " + userId + " reacted to message " + messageId + " with " + emoji);
        return true;
    }
    
    /**
     * Remove reaction from a message
     */
    public boolean removeReaction(String messageId, String userId) {
        Message message = messages.get(messageId);
        if (message == null) {
            return false;
        }
        
        message.removeReaction(userId);
        System.out.println("User " + userId + " removed reaction from message " + messageId);
        return true;
    }
    
    /**
     * Mark message as read by user
     */
    public boolean markAsRead(String messageId, String userId) {
        Message message = messages.get(messageId);
        if (message == null) {
            return false;
        }
        
        // In a real system, you'd track read status per user
        if (userId.equals(message.getRecipientId()) || message.getRoomId() != null) {
            message.markAsRead();
            return true;
        }
        
        return false;
    }
    
    /**
     * Get unread message count for user
     */
    public int getUnreadCount(String userId) {
        List<String> messageIds = privateMessages.get(userId);
        if (messageIds == null) {
            return 0;
        }
        
        return (int) messageIds.stream()
                .map(messages::get)
                .filter(Objects::nonNull)
                .filter(msg -> msg.getRecipientId().equals(userId) && 
                             msg.getStatus() != Message.MessageStatus.READ)
                .count();
    }
    
    /**
     * Clean up old messages based on retention policy
     */
    public void cleanupOldMessages() {
        LocalDateTime cutoffDate = LocalDateTime.now().minusDays(messageRetentionDays);
        
        List<String> messagesToDelete = messages.values().stream()
                .filter(msg -> msg.getTimestamp().isBefore(cutoffDate))
                .map(Message::getMessageId)
                .collect(Collectors.toList());
        
        // Delete old messages
        for (String messageId : messagesToDelete) {
            deleteMessageCompletely(messageId);
        }
        
        if (!messagesToDelete.isEmpty()) {
            System.out.println("Cleaned up " + messagesToDelete.size() + " old messages");
        }
    }
    
    /**
     * Completely delete a message from all indexes
     */
    private void deleteMessageCompletely(String messageId) {
        Message message = messages.get(messageId);
        if (message == null) {
            return;
        }
        
        // Remove from all indexes
        userMessages.getOrDefault(message.getSender().getUserId(), new ArrayList<>()).remove(messageId);
        
        if (message.isRoomMessage() && message.getRoomId() != null) {
            roomMessages.getOrDefault(message.getRoomId(), new ArrayList<>()).remove(messageId);
        } else if (message.isPrivateMessage() && message.getRecipientId() != null) {
            privateMessages.getOrDefault(message.getSender().getUserId(), new ArrayList<>()).remove(messageId);
            privateMessages.getOrDefault(message.getRecipientId(), new ArrayList<>()).remove(messageId);
        }
        
        if (message.getThreadId() != null) {
            threads.getOrDefault(message.getThreadId(), new ArrayList<>()).remove(messageId);
        }
        
        // Remove from main storage
        messages.remove(messageId);
    }
    
    /**
     * Get all messages (for admin/debugging)
     */
    public List<Message> getAllMessages() {
        return new ArrayList<>(messages.values());
    }
    
    /**
     * Get message service statistics
     */
    public Map<String, Object> getStatistics() {
        Map<String, Object> stats = new HashMap<>();
        stats.put("totalMessages", messages.size());
        stats.put("totalMessagesSent", totalMessagesSent);
        stats.put("totalMessagesDelivered", totalMessagesDelivered);
        stats.put("totalMessagesFailed", totalMessagesFailed);
        stats.put("pendingMessages", pendingMessages.size());
        stats.put("failedMessages", failedMessages.size());
        stats.put("threadsCount", threads.size());
        stats.put("roomsWithMessages", roomMessages.size());
        stats.put("usersWithMessages", userMessages.size());
        
        return stats;
    }
    
    @Override
    public String toString() {
        return "MessageService{messages=" + messages.size() + ", pending=" + pendingMessages.size() + "}";
    }
} 